File: src-tauri/src/deadlock.rs
use crate::error::AppError;
use std::path::{Path, PathBuf};

/// Known Steam library locations to search
fn get_steam_library_paths() -> Vec<PathBuf> {
    let mut paths = Vec::new();

    #[cfg(target_os = "linux")]
    {
        if let Some(home) = dirs::home_dir() {
            // Common Steam locations on Linux
            paths.push(home.join(".steam/steam/steamapps/common"));
            paths.push(home.join(".local/share/Steam/steamapps/common"));
            paths.push(home.join(".var/app/com.valvesoftware.Steam/.steam/steam/steamapps/common")); // Flatpak
        }
    }

    #[cfg(target_os = "windows")]
    {
        // Common Steam locations on Windows
        paths.push(PathBuf::from(r"C:\Program Files (x86)\Steam\steamapps\common"));
        paths.push(PathBuf::from(r"C:\Program Files\Steam\steamapps\common"));
        paths.push(PathBuf::from(r"D:\Steam\steamapps\common"));
        paths.push(PathBuf::from(r"D:\SteamLibrary\steamapps\common"));
    }

    #[cfg(target_os = "macos")]
    {
        if let Some(home) = dirs::home_dir() {
            paths.push(home.join("Library/Application Support/Steam/steamapps/common"));
        }
    }

    paths
}

/// Try to auto-detect Deadlock installation
pub fn detect_deadlock_path() -> Option<PathBuf> {
    for library_path in get_steam_library_paths() {
        let deadlock_path = library_path.join("Deadlock");
        if is_valid_deadlock_path(&deadlock_path) {
            return Some(deadlock_path);
        }
    }
    None
}

/// Validate that a path is a valid Deadlock installation
pub fn is_valid_deadlock_path(path: &Path) -> bool {
    // Check for expected game structure
    let game_dir = path.join("game");
    let citadel_dir = game_dir.join("citadel");

    game_dir.exists() && citadel_dir.exists()
}

/// Get the addons folder path, creating it if necessary
pub fn get_addons_path(deadlock_path: &Path) -> Result<PathBuf, AppError> {
    let addons_path = deadlock_path.join("game/citadel/addons");

    if !addons_path.exists() {
        std::fs::create_dir_all(&addons_path)?;
    }

    Ok(addons_path)
}

/// Get the disabled mods folder path, creating it if necessary
pub fn get_disabled_path(deadlock_path: &Path) -> Result<PathBuf, AppError> {
    let disabled_path = deadlock_path.join("game/citadel/addons/.disabled");

    if !disabled_path.exists() {
        std::fs::create_dir_all(&disabled_path)?;
    }

    Ok(disabled_path)
}

/// Get the gameinfo.gi file path
pub fn get_gameinfo_path(deadlock_path: &Path) -> PathBuf {
    deadlock_path.join("game/citadel/gameinfo.gi")
}

File: electron/main/services/deadlock.ts
import { existsSync, mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';

/**
 * Get platform-specific Steam library paths to search for Deadlock
 */
function getSteamLibraryPaths(): string[] {
    const paths: string[] = [];
    const home = homedir();

    if (process.platform === 'linux') {
        // Common Steam locations on Linux
        paths.push(join(home, '.steam/steam/steamapps/common'));
        paths.push(join(home, '.local/share/Steam/steamapps/common'));
        // Flatpak
        paths.push(join(home, '.var/app/com.valvesoftware.Steam/.steam/steam/steamapps/common'));
    } else if (process.platform === 'win32') {
        // Common Steam locations on Windows
        paths.push('C:\\Program Files (x86)\\Steam\\steamapps\\common');
        paths.push('C:\\Program Files\\Steam\\steamapps\\common');
        paths.push('D:\\Steam\\steamapps\\common');
        paths.push('D:\\SteamLibrary\\steamapps\\common');
    } else if (process.platform === 'darwin') {
        // macOS
        paths.push(join(home, 'Library/Application Support/Steam/steamapps/common'));
    }

    return paths;
}

/**
 * Check if a path is a valid Deadlock installation
 */
export function isValidDeadlockPath(path: string): boolean {
    const gameDir = join(path, 'game');
    const citadelDir = join(gameDir, 'citadel');
    return existsSync(gameDir) && existsSync(citadelDir);
}

/**
 * Auto-detect Deadlock installation path
 */
export function detectDeadlockPath(): string | null {
    const paths = getSteamLibraryPaths();
    console.log('[detectDeadlockPath] Searching paths:', paths);

    for (const libraryPath of paths) {
        const deadlockPath = join(libraryPath, 'Deadlock');
        console.log('[detectDeadlockPath] Checking:', deadlockPath);
        if (isValidDeadlockPath(deadlockPath)) {
            console.log('[detectDeadlockPath] FOUND:', deadlockPath);
            return deadlockPath;
        }
    }
    console.log('[detectDeadlockPath] Not found in any location');
    return null;
}

/**
 * Get the addons folder path, creating it if necessary
 */
export function getAddonsPath(deadlockPath: string): string {
    const addonsPath = join(deadlockPath, 'game', 'citadel', 'addons');

    if (!existsSync(addonsPath)) {
        mkdirSync(addonsPath, { recursive: true });
    }

    return addonsPath;
}

/**
 * Get the disabled mods folder path, creating it if necessary
 */
export function getDisabledPath(deadlockPath: string): string {
    const disabledPath = join(deadlockPath, 'game', 'citadel', 'addons', '.disabled');

    if (!existsSync(disabledPath)) {
        mkdirSync(disabledPath, { recursive: true });
    }

    return disabledPath;
}

/**
 * Get the gameinfo.gi file path
 */
export function getGameinfoPath(deadlockPath: string): string {
    return join(deadlockPath, 'game', 'citadel', 'gameinfo.gi');
}

/**
 * Get the citadel directory path
 */
export function getCitadelPath(deadlockPath: string): string {
    return join(deadlockPath, 'game', 'citadel');
}
